File: Data Structures & Algorithms/combination-target-sum/submission-3.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> curset;
        vector<vector<int>> comb;
        DFS(0 ,0 ,  target , nums , curset , comb);
        return comb;
    }

private:
    void DFS(int i ,int sum, int target , vector<int>& nums , vector<int>& curset , vector<vector<int>>& comb){
        if ( sum == target){
            comb.push_back(curset);
            return;
        }
        if ( i >= nums.size() || sum > target){
            return;
        }
        curset.push_back(nums[i]);
        DFS(i , nums[i] + sum , target , nums, curset,comb);
        curset.pop_back();

        DFS(i+1 , sum , target , nums, curset,comb);

    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        comb = []
        nums.sort()
        def backtrack(i , total):
            if total == target:
                res.append(comb.copy())
                return

            if total > target or i >= len(nums):
                return

            comb.append(nums[i])
            backtrack(i+1 , total + nums[i])
            comb.pop()

            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i+=1

            backtrack(i+1 , total)


        backtrack(0,0)

        return res